### PR TITLE
[Flink 16024][Connector][JDBC] Support FilterPushdown

### PR DESCRIPTION
**Pull request by qingwei91**
## What is the purpose of the change

Implement Filter Pushdown in JDBC Connector Source, this is an optimization that will avoid scanning the whole table into Flink when we only need a subset of the table.

## Brief change log

* Implement a SQL Visitor that converts Flink SQL AST to JDBC SQL String
* Implement FilterPushdown in JDBCDynamicSource using the Visitor
* Add tests for both Visitor and JDBCDynamicSource

## Verifying this change

This change added tests and can be verified as follows:

* Added Integrated Test in JdbcDynamicTableSourceITCase to make sure there is no regression on SQL Filtering
* Added JdbcFilterPushdownVisitorTest to verify we are generating SQL correctly

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (qingwei91) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:157]: This should be dialect specific so that it can be specialized to specific SQL syntax.

I noticed we removed all default implementation from this interface, should I do the same for this method?

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java
Comment (qingwei91) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:59]: Support Filter Pushdown in JDBC source

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java
Comment (qingwei91) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:235]: Core implementation of this change:

We traverse the ResolvedExpression, and produce a String if we know how to push it to SQL Database, returning None if we cannot handle it.

The unhandled expression will be kept as Flink SQL and run in the job

**Comment by MartijnVisser** [comment]
@qingwei91 Thanks for the PR. Can you make sure that your PR title and especially your commit messages are confirming to the Code Contribution guide? https://flink.apache.org/contributing/contribute-code.html

For example, your commit message should start with `[FLINK-16024][Connector][JDBC]` in this case.

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java
Comment (JingGe) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java]: It is not allowed to change public interface. Please check [FLIP-196: Source API stability guarantees](https://cwiki.apache.org/confluence/display/FLINK/FLIP-196%3A+Source+API+stability+guarantees) for more information.

**Review by JingGe — CHANGES_REQUESTED** [review]
Thanks @qingwei91 for driving this PR.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java]: Hi @JingGe, thanks for the comment.

This is a file I newly added, and in my latest commit, I marked it as PublicEvolving: https://github.com/apache/flink/pull/20140/commits/6c97da0eb0d27774ed9954fbabed7b514ac8ce02

Do you mean we cannot expose new API? 
For instance, I added this here: https://github.com/apache/flink/pull/20140/files#diff-ae60653ffe2ac890a3c1b01da41405bcc4e6913949176c36edc009df5090c38fR157, which adds a new method and a new return type to `JdbcDialect` which is PublicEvolving, because filter pushdown might differ across JDBC dialect. Is this approach a problem?

Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java]: Hi @JingGe did you get a chance to check on this?

Comment (JingGe) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java]: Hi @qingwei91 Could you please explain why is a public API required? Does @Internal work too? Thanks.

Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java]: Hi @JingGe , I believe it has to be PublicEvolving because this class is returned by a public method in JdbcDialect, which is a PublicEvolving class.

I will sort out the commit, sorry for the confusion

Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownVisitor.java]: I've squash the commits now and we have a green build.

Can anyone of you review this again? Thank you!

**Comment by libenchao** [comment]
@qingwei91 Thanks for your contribution. For the current design, I see you use `ValueLiteralExpression#toString()` to generate the string for literals. This may work for some cases, but not for all cases IMHO.
Consider following cases:
1. `ValueLiteralExpression#toString()` uses Flink dialect to represent string, and it hard-coded the quote with `'`. However, in many DBMS, `'` is not the only choice. 
2. `ValueLiteralExpression#toString()` only handles special character `'` using escape. In many DBMS, they need more special character handing, e.g. [mysql driver](https://github.com/mysql/mysql-connector-j/blob/release/8.0/src/main/protocol-impl/java/com/mysql/cj/protocol/a/StringValueEncoder.java#L72)
3. For other types, e.g. `TIMESTAMP`, `DATE`, `TIME`, `INTERVAL` and so on, they may suffers from this too, because we cannot assume that all DB dialects handle them in the same way.

Another more general way to handle this is to use `PrepareStatement.setXXX` just like we already did in `TableSimpleStatementExecutor` and `JdbcRowDataLookupFunction`. WDYT?

**Comment by qingwei91** [comment]
Hi @libenchao , thanks for the review!

Thanks for pointing out the flaw, you're right. 🙇 

On your recommended approach, I believe `JdbcFilterPushdownVisitor` needs to produce strings or some equivalent data, so that it can be used by `JdbcDynamicTableSource::getScanRuntimeProvider`, how should I make use of PrepareStatement here? Maybe I am missing something?

I think you're pointing out a fundamental issue with this PR, SQL statement generation has to be dialect-specific, and me trying to provide a default implementation might be a lost cause here.

If we cannot go down the prepared statement route, I can think of 2 ideas:

1. Implement dialect-specific method that converts `ValueLiteralExpression` to SQL string literal, and have `JdbcFilterPushdownVisitor` to make use of it. We don't have to support all dialect in 1 go, and we can simply fallback to in-memory filter when its not supported.
2. Let `JdbcFilterPushdownVisitor` be dialect-specific, then every implementation will need to deal with dialect specific thing including literal stringification. This is similar to current approach, the only difference is that currently this PR provides a default implementation, where this option will stop doing it. My implementation is tested and used in Production for SQL Server. Maybe I can rename it and make it specific to SQL Server jdbc dialect? (SQL Server dialect is still opened #20235), so this will have a dependency to it. Likewise, we can fallback to in-memory filtering when dialect without an implementation.

I think option 1 is less code, but probably more fiddly and easier to break. Option 2 is likely gonna be more code, but the separation is cleaner and less likely to break.

Let me know your thoughts. 😄

**Comment by libenchao** [comment]
>  I believe JdbcFilterPushdownVisitor needs to produce strings or some equivalent data, so that it can be used by JdbcDynamicTableSource::getScanRuntimeProvider, how should I make use of PrepareStatement here? Maybe I am missing something?

@qingwei91 Currently `JdbcRowDataInputFormat` already uses `PreparedStatement`, and 'scan.partition' is implemented using it, hence we can do it. `JdbcFilterPushdownVisitor` does not necessarily need to return `String`, it can  return anything we need.
We may need to rethink about the design, especially for the following points:
1. What functions should we support, e.g. `IN`, `BETWEEN`?
2. Are these functions supported in all JDBC dialects, and do they use the same operator? If not, how should we abstract it out to make it dialect specific?

**Comment by qingwei91** [comment]
Hi @libenchao, oh I see, I didnt spot that `JdbcRowDataInputFormat` is using PreparedStatement under the hood. 

Thanks for pointing it out, this will be larger change than I expect, I will give it a go.

> 1. What functions should we support, e.g. IN, BETWEEN?

I think we can tackle this incrementally? I believe IN is supported out of the box, because Flink compile IN into multiple X=Y condition chained together by OR, I never looked into BETWEEN though

> 2. Are these functions supported in all JDBC dialects, and do they use the same operator? If not, how should we abstract it out to make it dialect specific?

I think ultimate we need to allow dialect specificness, right now I design it such that the query generator (ie. JdbcFilterPushdownVisitor) is part of JDBCDialect, so each dialect can provide their own instance to deal with it. Do you think this design is okay? Or is there a better way?

**Comment by libenchao** [comment]
> I believe IN is supported out of the box, because Flink compile IN into multiple X=Y condition chained together by OR

Not exactly, we have a threshold (default 4): https://github.com/apache/flink/blob/208f08b406a7fd48890cda16d317a30ee892a2e7/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala#L47

> I think we can tackle this incrementally? 

I agree, we can start from some common functions, such as `=`, `<>`, `<`, `>`, `<=`, `>=`, `IS NULL`, `IS NOT NULL`, `IN`, and leave others to the future.

**Comment by HeChuanXUPT** [comment]
Wow, Thanks, It works for me~

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (grzegorz8) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:158]: Timestamps might be tricky. What if one database accepts `yyyy-MM-dd'T'HH:mm:ss.SSSZ` while other one requires `yyyy-MM-dd HH:mm:ss.SSSZ` (note there is no `'T'`)? The format might differ per dialect.

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
Hi @libenchao I finally got around and implemented your suggestion. Do you mind to take a look again?

I didn't implement additional operators like IN and BETWEEN because I am quite busy of late. Given they can be implemented later incrementally, and its not going to break anything, I wonder if we can defer those to a different ticket?

**Comment by libenchao** [comment]
@qingwei91 Thanks for the updating, I'll review this PR in a few days. 

> I didn't implement additional operators like IN and BETWEEN because I am quite busy of late. Given they can be implemented later incrementally, and its not going to break anything, I wonder if we can defer those to a different ticket?

I'm ok with this.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:69]: `log` -> `LOG`

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:180]: It's not necessary to do `if (log.isDebugEnabled())`?

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:213]: Make a deep copy for these two params cause they are mutable.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:64]: These two fields can both be `private`?

Comment (libenchao) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:64]: And we usually use upper case for static field names, also make fields `final` as much as possible.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:87]: How about adding `<>` too since it's also a simple binary operator.

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitorITCase.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitorITCase.java:65]: Why do you name this an `ITCase` instead of normal UT?

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java:268]: Could you also add a test in `JdbcTablePlanTest` like `JdbcTablePlanTest#testLimitPushDown`?

**Review by libenchao — COMMENTED** [review]
@qingwei91 Thanks for your update and sorry for the delay. The PR is in a good shape now, I've left some minor comments.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:87]: Will do 👍

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java:268]: Sure, will do

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitorITCase.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitorITCase.java:65]: My bad, it is a mistake

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
Hi @libenchao , thank you very much for your review 👍 

I've addressed all of your concern. On the TablePlanTest, do you mind to check if that's how it supposed to work? I don't think I understand internal good enough to judge

File: flink-connectors/flink-connector-jdbc/src/test/resources/org/apache/flink/connector/jdbc/table/JdbcTablePlanTest.xml
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/test/resources/org/apache/flink/connector/jdbc/table/JdbcTablePlanTest.xml:67]: why the condition still exists in the Calc?

**Review by libenchao — COMMENTED** [review]
One small tip, do not rebase/force-pushing before the reviewer asks you because this will make the reviewer hard to do incremental review.

File: flink-connectors/flink-connector-jdbc/src/test/resources/org/apache/flink/connector/jdbc/table/JdbcTablePlanTest.xml
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/test/resources/org/apache/flink/connector/jdbc/table/JdbcTablePlanTest.xml]: I figure out, its because the column type Time wasn't supported, I've added support here: https://github.com/apache/flink/pull/20140/commits/74fe5eed158482920a9ec36ac8a7aad549ee94fa

Thanks for calling it out.

**Comment by qingwei91** [comment]
> One small tip, do not rebase/force-pushing before the reviewer asks you because this will make the reviewer hard to do incremental review.

Sorry, my bad, I was advised to squashed in prev PR, but of course that should only be done right before merging.

**Comment by qingwei91** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java]: Great idea! I didnt discover this typeRoot thingy earlier, thanks for the suggestion.

https://github.com/apache/flink/pull/20140/commits/d9b14572a0661f145cab5344a8c9b2076b75b049

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

**Comment by qingwei91** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:102]: 1. could you sort out the case branches, similar to `LogicalTypeRoot`
2. there are still some left types, such as `CHAR`, `SMALLINT`, `TINYINT`, I would suggest that we add these types as much as possible, unless we cannot for now.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:49]: `final`

**Review by libenchao — COMMENTED** [review]
@qingwei91 Thanks for the update. Please do not use 'squash' or 'force-push' unless you must or the reviewer asks you. (I go through all the codes again, and left several minor comment)

**Comment by qingwei91** [comment]
@libenchao 

> Thanks for the update. Please do not use 'squash' or 'force-push' unless you must or the reviewer asks you. (I go through all the codes again, and left several minor comment)

Sorry, in the last commit, I thought to bundle all changes I've made since your last review into https://github.com/apache/flink/pull/20140/commits/55d5227fe040890f0c89e366a5ae19c6c75c1d21, so that you can review just that without noise. I wont do it again.

I will go through the comments and fix them

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:102]: Yes, will do

**Comment by qingwei91** [comment]
Hi @libenchao , this is the new commit I added to address your comment.

I also added support IS NULL and IS NOT NULL as these 2 are quite common

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (libenchao) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:103]:
```suggestion
            String operator, ResolvedExpression operand, boolean operandOnLeft) {
```

**Review by libenchao — COMMENTED** [review]
> I also added support IS NULL and IS NOT NULL as these 2 are quite common

This is great! 
The PR LGTM with only one minor comment, after fixing that, I'll merge this.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java]: Thanks, committed 👍

**Review by libenchao — APPROVED** [review]
LGTM, merging. @qingwei91 Thanks for your contribution and the patience and consistence during the review.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java
Comment (grzegorz8) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:57]: @qingwei91 - Is there any reason why LIKE was not added here? It seems to be as easy as other binary operators below.
I will gladly add LIKE but I just want to ensure I didn't overlook something tricky.

Comment (qingwei91) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcFilterPushdownPreparedStatementVisitor.java:57]: Hi @grzegorz8 , I think we simply missed it, thanks for noticing this.

**Pull request closed without merging**
